src/tracer.cc

```diff
@@ -279,34 +312,75 @@ TraceManager::SampleTrace(const std::string& model_name)
 
 TraceManager::Trace::~Trace()
 {
-  // Write trace now
-  setting_->WriteTrace(streams_);
+  if (setting_->mode_ == TRITONSERVER_TRACE_MODE_TRITON) {
+    // Write trace now
+    setting_->WriteTrace(streams_);
+#ifndef _WIN32
+  } else if (setting_->mode_ == TRITONSERVER_TRACE_MODE_OPENTELEMETRY) {
+    this->CleanTracer();
+#endif
+  }
 }
 
 void
 TraceManager::Trace::CaptureTimestamp(
     const std::string& name, uint64_t timestamp_ns)
 {
   if (setting_->level_ & TRITONSERVER_TRACE_LEVEL_TIMESTAMPS) {
-    std::lock_guard<std::mutex> lk(mtx_);
-    std::stringstream* ss = nullptr;
-    {
-      if (streams_.find(trace_id_) == streams_.end()) {
-        std::unique_ptr<std::stringstream> stream(new std::stringstream());
-        ss = stream.get();
-        streams_.emplace(trace_id_, std::move(stream));
-      } else {
-        ss = streams_[trace_id_].get();
-        // If the string stream is not newly created, add "," as there is
-        // already content in the string stream
-        *ss << ",";
+    if (setting_->mode_ == TRITONSERVER_TRACE_MODE_TRITON) {
+      std::lock_guard<std::mutex> lk(mtx_);
+      std::stringstream* ss = nullptr;
+      {
+        if (streams_.find(trace_id_) == streams_.end()) {
+          std::unique_ptr<std::stringstream> stream(new std::stringstream());
+          ss = stream.get();
+          streams_.emplace(trace_id_, std::move(stream));
+        } else {
+          ss = streams_[trace_id_].get();
+          // If the string stream is not newly created, add "," as there is
+          // already content in the string stream
+          *ss << ",";
+        }
       }
+      *ss << "{\"id\":" << trace_id_ << ",\"timestamps\":["
+          << "{\"name\":\"" << name << "\",\"ns\":" << timestamp_ns << "}]}";
+#ifndef _WIN32
+    } else if (setting_->mode_ == TRITONSERVER_TRACE_MODE_OPENTELEMETRY) {
+      otel_common::SystemTimestamp otel_timestamp{
+          (time_offset_ + std::chrono::nanoseconds{timestamp_ns})};
+      if (trace_span_ == nullptr) {
+        this->InitSpan(otel_timestamp);
+      }
+      trace_span_->AddEvent(
+          name, otel_timestamp, {{"triton.steady_timestamp_ns", timestamp_ns}});
+#endif
```

Comment:
Should write a else block for these conditionals. And generate a warning message for unsupported trace_mode.

Reply:
updated